### PR TITLE
Add S3 API execution trace to MinioException

### PR DESCRIPTION
**Pull request by balamurugana**
This is useful to debug issues further and to find root cause.

### CODE REVIEW

File: api/src/main/java/io/minio/errors/BucketPolicyTooLargeException.java [hunk 1]
```diff
@@ -16,23 +16,10 @@
 
 package io.minio.errors;
 
-/** Thrown to indicate that given bucket name is not valid. */
+/** Thrown to indicate that received bucket policy is larger than 12KiB size. */
```
Comment (kannappanr) [inline api/src/main/java/io/minio/errors/BucketPolicyTooLargeException.java:19]: https://docs.aws.amazon.com/AmazonS3/latest/dev/example-bucket-policies.html

according to AWS S3 and MinIO the max size is 20 KiB

Comment (balamurugana) [reply api/src/main/java/io/minio/errors/BucketPolicyTooLargeException.java:19]: Looks like it is increased from 12KB to 20KB

**Review by kannappanr — DISMISSED** [review]
LGTM

**Comment by kannappanr** [comment]
@balamurugana PTAL build failure

**Review by kannappanr — APPROVED** [review]

File: api/src/main/java/io/minio/MinioClient.java [hunk 2]
```diff
@@ -618,25 +628,28 @@ protected Response execute(
...
+            .toString()
+            .replaceAll("Signature=([0-9a-f]+)", "Signature=*REDACTED*")
+            .replaceAll("Credential=([^/]+)", "Credential=*REDACTED*"));
+    if (traceRequestBody) {
+      traceBuilder.append("\n").append(new String((byte[]) body, StandardCharsets.UTF_8));
     }
 
+    PrintWriter traceStream = this.traceStream;
+    if (traceStream != null) traceStream.println(traceBuilder.toString());
+    traceBuilder.append("\n");
```
[the remark this answers is not in the document]
Comment (balamurugana) [reply api/src/main/java/io/minio/MinioClient.java:651]: Done

File: api/src/main/java/io/minio/MinioClient.java [hunk 3]
```diff
@@ -645,15 +658,27 @@ protected Response execute(
...
-      this.traceStream.println(response.headers());
-    }
+    String trace =
+        response.protocol().toString().toUpperCase(Locale.US)
+            + " "
+            + response.code()
+            + "\n"
+            + response.headers();
+    traceBuilder.append(trace).append("\n");
+    if (traceStream != null) traceStream.println(trace);
```
[the remark this answers is not in the document]
Comment (balamurugana) [reply api/src/main/java/io/minio/MinioClient.java:668]: I think it is not useful for the one liner.

File: api/src/main/java/io/minio/MinioClient.java [hunk 4]
```diff
@@ -229,6 +230,15 @@
...
   }
 
+  private static final Set<String> TRACE_QUERY_PARAMS = new HashSet<>();
+
+  static {
+    TRACE_QUERY_PARAMS.add("retention");
+    TRACE_QUERY_PARAMS.add("legal-hold");
+    TRACE_QUERY_PARAMS.add("tagging");
+    TRACE_QUERY_PARAMS.add(UPLOAD_ID);
+  }
```
[the remark this answers is not in the document]
Comment (anjalshireesh) [reply api/src/main/java/io/minio/MinioClient.java:240]: May not be a bad idea to change only in this place as it's a new code being added in this PR and later fix the other instances.

File: api/src/main/java/io/minio/MinioClient.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (anjalshireesh) [reply api/src/main/java/io/minio/MinioClient.java:675]: Sometimes it makes sense to have even single-statement methods if it improves readability. In this instance, there are three statements and not one.

**Review by anjalshireesh — APPROVED** [review]
My comments are minor, so approving the PR anyway.

**Review by sinhaashish — APPROVED** [review]

**Pull request merged**